code/lib/manager-api/src/index.tsx

```diff
@@ -419,36 +419,41 @@ const addonStateCache: {
 // shared state
 export function useSharedState<S>(stateId: string, defaultState?: S) {
   const api = useStorybookApi();
-  const existingState = api.getAddonState<S>(stateId);
+  const existingState = api.getAddonState<S>(stateId) || addonStateCache[stateId];
   const state = orDefault<S>(
     existingState,
     addonStateCache[stateId] ? addonStateCache[stateId] : defaultState
   );
-  const setState = (s: S | API_StateMerger<S>, options?: Options) => {
-    // set only after the stories are loaded
-    if (addonStateCache[stateId]) {
+
+  if (api.getAddonState(stateId) !== state) {
+    api.setAddonState<S>(stateId, state).then((s) => {
       addonStateCache[stateId] = s;
-    }
-    api.setAddonState<S>(stateId, s, options);
+    });
+  }
+
+  const setState = (s: S | API_StateMerger<S>, options?: Options) => {
+    const result = api.setAddonState<S>(stateId, s, options);
+    addonStateCache[stateId] = result;
+    return result;
   };
   const allListeners = useMemo(() => {
     const stateChangeHandlers = {
-      [`${SHARED_STATE_CHANGED}-client-${stateId}`]: (s: S) => setState(s),
-      [`${SHARED_STATE_SET}-client-${stateId}`]: (s: S) => setState(s),
+      [`${SHARED_STATE_CHANGED}-client-${stateId}`]: setState,
+      [`${SHARED_STATE_SET}-client-${stateId}`]: setState,
     };
     const stateInitializationHandlers = {
-      [SET_STORIES]: () => {
+      [SET_STORIES]: async () => {
```

Comment:
What's going on with this code @ndelangen? Given we don't fire `SET_STORIES` for SSv7, it seems weird to have code that relies on it.

Reply:
The code handling this event was already there, I just made a change to ensure `await` setting state before emitting the event.

`useSharedState` can be used with SSv6 as well.

Reply:
I know I guess I just meant while you were looking at it maybe we should consider removing it given it doesn't run in ssv7 and I guess things still work?

Reply:
@tmeasday I'll leave the code alone for now, and we'll remove SSv6 hopefully in 8.0 🤞